### PR TITLE
Dataset Creation Changes

### PR DESCRIPTION
**Pull request by AbhiDhir**
This PR adds the core functionality for dataset creation which dataset plugins can use. Plugins will receive a CreateInput object and are expected to return a CreateOutput object. The core handles ensuring image sets that the user requires are downloaded and synched with S3, and paths to where datasets are meant to be created are clear as well. Once the plugin has finished running, the core also handles uploading of the dataset to S3 and removal of any unnecessary local files. Below I have explanations of some of the higher-level interfaces/classes, but they are documented in more detail in the code.

Usage:
`ravenml data create -c <config file> <plugin>`

Interfaces:
CreateInput (What the plugin is given):
- Has information that the plugins can use
  - Includes (but not limited to) dataset metadata, paths to data directory and dataset, dataset config (e.g. kfolds, test_percent)

CreateOutput (What the plugin is expected to return):
 - Currently empty (as the plugin is expected to write their dataset into the directory specified by the CreateInput's `dataset_path`)

The` DatasetWriter` class contains much of the functionality needed to create a dataset. It serves as a guide to what methods can be called in dataset creation (loading image_ids, filtering, constructing dataset objects, writing files, etc.). `DefaultDatasetWriter` has much of this functionality baked into it, but is mostly modular where plugins can choose to override methods as they choose. The only input needed is the `CreateInput` as well as a dictionary of `associated_files` for usage in image_id discovery and copying of files.

RMLCache directory layout:
~/.ravenML
│── sample_dataset_plugin
│   └── datasets (Default location for datasets if no path is specified)
│── imagesets (Imageset cache, synched with s3)
    └── imageset_1
    └── imageset_2

 Misc Changes/Features:
- `aws s3 sync` is now in use for downloading/uploading files
- all `object_detection` dependencies have been replaced
- local imagesets can be used for dataset creation

NOTE:
- There are still a few areas in the code that are subject to review/discussion. I will comment on them individually.

### CODE REVIEW

File: ravenml/data/write_dataset.py [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+            which will be later used by the plugin to create their dataset.
+
+            If overloaded, method is expected to copy all files the plugin needs into the provided 'temp_dir'.
+
+        Variables Needed:
+            image_ids (list): needed to find what needs to be copied (provided by 'load_image_ids'/'interactive_filter')
+            temp_dir (Path): needed to know where to copy to (provided by 'create' input)
+            associated_files (dict): needed to know what files need to be copied (provided by plugin)
+        """
+        copy_data_locally(self.image_ids, self.temp_dir, self.associated_files)
```
Comment (AbhiDhir) [inline ravenml/data/write_dataset.py]: @cSchubes @kevin3-black This is where data is copied into a temp directory. Pros of changing to sending plugins paths to the actual data is that it wastes less resources. Cons are that it is possible for plugins to screw up the data (although only if they add files to the imageset caches since `s3 sync` will take care of changed or deleted files). I don't feel too strongly on the topic, but if it does save a significant amount of time for larger datasets it may be worth it since it doesn't seem too difficult to implement.

File: ravenml/data/write_dataset.py [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+        test_subset, dev_subset = split_data(obj_list, test_percent=self.test_percent)
+
+        # Test subset
+        test_path = dataset_path / 'test'
+        os.mkdir(test_path)
+        test_image_ids = [(self.temp_dir, obj['image_id']) for obj in test_subset]
+        copy_data_locally(test_image_ids, test_path, self.associated_files)
+
+        dev_path = dataset_path / 'splits'
+
```
Comment (AbhiDhir) [inline ravenml/data/write_dataset.py:268]: There was discussion before on what the standard internal structure of datasets should look like. Currently kfolds is not built into core so the training data is found in dataset_path/splits/complete.

File: ravenml/data/helpers.py [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+from random import shuffle
+from queue import Queue
+from threading import Thread
+from pathlib import Path
+from colorama import Fore
+from ravenml.utils.question import cli_spinner, user_selects, user_confirms, user_input
+from ravenml.utils.config import get_config
+
+def default_filter(tags_df, filter_metadata):
+    """Method leads user through interactive filtering through image_ids based on
```
Comment (AbhiDhir) [inline ravenml/data/helpers.py:15]: Additional parameters (filtering by the amount of images needed as well as by tags) are being worked on, will most likely be in a separate pr.

File: ravenml/data/interfaces.py [hunk 4]
```diff
@@ -6,23 +6,179 @@
...
 ### CONSTANTS ###
 # these should be used in all possible situations to protect us
 # in case they change in the future
 STANDARD_DIR = 'standard'
 FOLD_DIR_PREFIX = 'fold_'
 TEST_DIR = 'test'
-
-# TODO add necessary functionality to this class as needed
+METADATA_PREFIX = 'meta_'
+TEMP_DIR_PREFIX = 'datasets/temp'
```
Comment (kvablack) [inline ravenml/data/interfaces.py]: Looks like STANDARD_DIR, TEST_DIR, and METADATA_PREFIX aren't used anywhere. I'm guessing TEST_DIR should be used somewhere (and isn't), and METADATA_PREFIX should be removed because it's plugin-specific.

File: ravenml/data/write_dataset.py [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+                tags associated with each image_id
+            image_ids (list): list of tuples containing a path to an imageset
+                and an image_id in that imageset
+            filter_metadata (dict): holds the groups of different subsets of
+                image_ids
+        """
+
+        self.associated_files = kwargs['associated_files']
+        if not self.associated_files.get('metadata'):
+            raise Exception("Associated files must contain a 'metadata' key with a corresponding prefix-suffix pair")
```
Comment (kvablack) [inline ravenml/data/write_dataset.py]: Shouldn't the whole concept of `associated_files` go in DefaultDatasetWriter, since DatasetWriter is just an interface? Also, this validation should go in `load_image_ids` in the DefaultDatasetWriter, because if a plugin wants to override `load_image_ids` then they shouldn't be required to have a 'metadata' key.

EDIT: If you do end up getting rid of `load_data` and making the changes I suggested to `write_dataset`, then `associated_files` could actually be moved out of the constructor completely. The default implementation of `load_image_ids` only needs a metadata prefix, which it could get passed directly. The method to create the test set would be the only place that needs the full `associated_files` list.

File: ravenml/data/write_dataset.py [hunk 6]
```diff
@@ -0,0 +1,363 @@
+import os, shutil, time, json
+import pandas as pd
+from pathlib import Path
+from datetime import datetime
+from ravenml.data.interfaces import CreateInput
+from ravenml.utils.question import cli_spinner, cli_spinner_wrapper
+from ravenml.utils.config import get_config
+from ravenml.data.helpers import default_filter, copy_data_locally, split_data, read_json_metadata
+
+class DecoratorSuperClass:
```
Comment (kvablack) [inline ravenml/data/write_dataset.py]: This is really awesome. However, it should probably go in a more generic utility file (perhaps the same place as `cli_spinner_wrapper`.

File: ravenml/data/write_dataset.py [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+        metadata_prefix = self.associated_files['metadata'][0]
+        metadata_suffix = self.associated_files['metadata'][1]
+        if metadata_suffix != '.json':
+            raise Exception("Currently non-json metadata files are not supported for the default loading of image ids")
+        
+        # Goes through each file in each imageset to search for metadata files
+        # metadata files are parsed for tags and filename is parsed for image_id 
+        for data_dir in self.imageset_paths:
+            for dir_entry in os.scandir(data_dir):
+                if not dir_entry.name.startswith(metadata_prefix):
```
Comment (kvablack) [inline ravenml/data/write_dataset.py]: Shouldn't this also check for metadata_suffix?

File: ravenml/data/write_dataset.py (hunk 1 above)
Comment (kvablack) [reply ravenml/data/write_dataset.py]: Yeah, I do think that in the long run getting rid of this copying is a good idea, especially since `s3 sync` automatically takes care of any accidental changes or deletions. However, as long as this works I'd say it's fine for now.

EDIT: Actually, if the implementation fairly easy, then it might be worthwhile to just put it in this PR. The path to the imageset is already part of the `image_ids` list, so it seems like it's already all set up to skip the copying and work with the data directly. Also, the whole point of recording the imageset that each image came from was to avoid image ID conflicts --- but if two imagesets have conflicting image IDs with the same prefix and suffix, then won't they still overwrite each other when copying into the same temporary directory?

Finally, getting rid of this extra method will further simplify the interface and make it easier to understand, which is always a plus. If a plugin wants to change the way they load their data, then they have one less method to worry about.

File: ravenml/data/helpers.py [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+def copy_data_locally(images: list, destination_dir: Path, associated_files: dict, num_threads=20):
+    """Copies files associated with provided image list into a destination 
+        directory locally
+    
+    Args:
+        images (list): list of tuples with paths to a local directory paired 
+            with an image id located in that directory
+        destination_dir (Path): directory where data will be copied to
+        associated_files (dict): expected to follow the format:
+            { 'filetype' (String): ('prefix', 'suffix') (tuple) }
```
Comment (kvablack) [inline ravenml/data/helpers.py]: Maybe document what "filetype" is supposed to be? Doesn't look like it's used for anything right now, so it could be removed.

File: ravenml/data/write_dataset.py [hunk 9]
```diff
@@ -0,0 +1,363 @@
...
+            that each image_id corresponds to a metadata file. Once a metadata file is found (based on the
+            metadata prefix-suffix tuple provided in 'associated_files') image_id is extracted from the file
+            name and the file is parsed to get tag information. Currently only json metadata files are
+            supported in this default implementation.
+
+            If overloaded, method is expected to set 'self.tags_df' if wanting to support 'interactive_filter'.
+            'self.tags_df' should be set to a pandas dataframe created from a dict with tuple pairs: 
+            (imageset_path (Path), image_id (String)) as keys and an array of True/False values for whether 
+            the image_id has each corresponding tag. 'self.image_ids' is also expected to be set to a list of 
+            image_ids if any other methods need to be used.
```
Comment (kvablack) [inline ravenml/data/write_dataset.py]: Perhaps document more clearly that `image_ids` is supposed to be a list of tuples with `imageset_path` as the first element and the actual ID as the second. I was confused for a bit when looking at other code because I was assuming that `image_ids` was just a list of image ID strings. It should maybe even be renamed, although admittedly I can't think of a better one.

File: ravenml/data/write_dataset.py [hunk 10]
```diff
@@ -0,0 +1,363 @@
...
+        __init__ (CreateInput): takes in CreateInput object and initializes variables
+            to be used by the rest of the methods and the plugins
+        load_image_ids (): gets all image_ids/tags from the supplied imagesets based on
+            finding metadata files
+        interactive_filter (): takes the current image_ids and allows the user to create
+            sets through interactive filtering using the image_ids
+        load_data (): copies all related files to the image_ids into a temp folder
+        construct_all (): plugin specific method to generate objects which will be used
+            in writing the dataset
+        write_dataset (obj_list): main driver for writing the dataset locally
```
Comment (kvablack) [inline ravenml/data/write_dataset.py]: Why do all the other methods rely on setting instance variables, but then `write_dataset` requires an argument? I figure may as well standardize it so that they all use the same paradigm.

File: ravenml/data/write_dataset.py [hunk 11]
```diff
@@ -0,0 +1,363 @@
...
+        dataset_path = self.dataset_path / self.dataset_name
+        print(dataset_path)
+
+        test_subset, dev_subset = split_data(obj_list, test_percent=self.test_percent)
+
+        # Test subset
+        test_path = dataset_path / 'test'
+        os.mkdir(test_path)
+        test_image_ids = [(self.temp_dir, obj['image_id']) for obj in test_subset]
+        copy_data_locally(test_image_ids, test_path, self.associated_files)
```
Comment (kvablack) [inline ravenml/data/write_dataset.py]: I think this default implementation of `write_dataset` could use a bit more work. It's so close to being completely agnostic to the exact way that the plugin constructs its data, but these two lines force the plugin to make every object a dictionary with an 'image_id' key. Since the only reason for this is for the test set, perhaps the creation of the test set should be split out into its own overridable method (just like `write_data`). The default implementation could rely on the objects having an `image_id` attribute, but the plugin could override it without having to override `write_dataset`.

File: ravenml/data/helpers.py [hunk 12]
```diff
@@ -0,0 +1,311 @@
...
+                    in that column header
+    """
+    result = pd.DataFrame()
+    for group in sets:
+        result = pd.concat((result, group), sort=False)
+    result = result.reset_index().drop_duplicates(
+        subset='index', keep='first').set_index('index')
+    return result
+
+def copy_data_locally(images: list, destination_dir: Path, associated_files: dict, num_threads=20):
```
Comment (kvablack) [inline ravenml/data/helpers.py]: This name is kind of a misnomer, since it doesn't have anything to do with "local" data anymore (I know this is a holdover from Jigsaw when the temp dir was always in the CWD). Perhaps `copy_associated_files` would be a better name.

File: ravenml/utils/question.py [hunk 13]
```diff
@@ -144,4 +144,21 @@ def cli_spinner(text, func, *args, **kwargs):
...
+        def wrapper(*args, **kwargs):
+            spinner = Spinner(text=text, text_color="magenta")
+            spinner.start()
+            try:
+                result = func(*args,**kwargs)
+            except Exception:
+                spinner.succeed(text=text + 'Failed.')
+                raise
+            spinner.succeed(text=text + 'Complete.')
+            return result
```
Comment (kvablack) [inline ravenml/utils/question.py]:
```suggestion
            return cli_spinner(text, func, *args, **kwargs)
```

**Review by kvablack — CHANGES_REQUESTED** [review]

File: ravenml/data/write_dataset.py (hunk 7 above)
Comment (AbhiDhir) [reply ravenml/data/write_dataset.py]: Right now this implementation only supports parsing json metadata files. I have a check to see if the `metadata_suffix` is json earlier. If it is not then an exception is raised, which is why it doesn't need to be checked in the loop.

Comment (kvablack) [reply ravenml/data/write_dataset.py]: What if there is a file called `meta_something.png`? I know this is an edge case, but it seems like the suffix should be checked just for completeness (as long as the method takes a suffix). I did mention in another comment though that you could also just change this method to explicitly take a metadata prefix as an argument, which I think would also be fine.

File: ravenml/data/helpers.py [hunk 14]
```diff
@@ -206,18 +206,18 @@ def join_sets(sets):
         subset='index', keep='first').set_index('index')
     return result
 
-def copy_data_locally(images: list, destination_dir: Path, associated_files: dict, num_threads=20):
+def copy_associated_files(images: list, destination_dir: Path, associated_files: dict, num_threads=20):
```
Comment (kvablack) [inline ravenml/data/helpers.py]: Type hint should be `associated_files: list`

File: ravenml/data/write_dataset.py [hunk 15]
```diff
@@ -342,6 +271,23 @@ def write_dataset(self, obj_list: list):
...
+
+            If overloaded, there are no expectations.
+
+        Args:
+            path (Path): Path to where data should be written
+            data (list): data that should be written
+            associated_files (list): decides what files are to be copied for the test set
+        """
+        os.mkdir(path)
+        test_image_ids = [(Path(os.path.dirname(obj['image_filepath'])), obj['image_id']) for obj in data]
```
Comment (kvablack) [inline ravenml/data/write_dataset.py]: Looking at the plugin code, `image_filepath` is the full path to the image. Shouldn't this be the path to the imageset directory instead?

Comment (kvablack) [inline ravenml/data/write_dataset.py]: Hmm this is much better, but I realized the reliance on the object having the imageset path and image ID still feels redundant when all that info is in `self.image_ids` already. Perhaps a solution could be having `construct_all` create a dictionary mapping image IDs to data objects rather than a plain list of data objects?

File: ravenml/data/write_dataset.py [hunk 16]
```diff
@@ -0,0 +1,309 @@
...
+        copy_associated_files(test_image_ids, path, associated_files)
+
+    def write_out_complete_set(self, path, data):
+        """Method is helper function for writing out dataset. Creates a 
+            'train' subdirectory and calls for 'write_data' to write
+            test_data and _train_data.
+
+            If overloaded, there are no expectations.
```
Comment (kvablack) [inline ravenml/data/write_dataset.py]: This is a nitpick, because people will generally know what you mean, but all instances of "overload" should technically say "override"

File: ravenml/data/write_dataset.py [hunk 17]
```diff
@@ -0,0 +1,309 @@
...
+    def __init__(self, create: CreateInput):
+        """Method calls DatasetWriter's initialization to get all variables it needs
+
+        Args:
+            create (CreateInput): what is passed to the plugin,
+                containing configuration information
+        """
+        super().__init__(create)
+
+    def write_data(self, objects, path, split_type, *args, **kwargs):
```
Comment (kvablack) [inline ravenml/data/write_dataset.py]: It would be nice if this method name was a bit more specific, it's kind of easy to get it mixed up with `write_dataset`. Ideally, it would indicate the kind of data it's writing, similar to `write_out_test_set`.

The problem is that our naming convention is pretty terrible, with the `test.record` and top-level holdout `test` directory being completely different things. At some point, we should really rename the `test.record` data to something like 'validation' or 'development' which are both commonly used. We also have to keep in mind that this method will eventually be used to write out the splits of k-fold cross-validation, so the naming convention has to make sense in that context as well.

However, this is a whole other discussion and out of the scope of this PR. For now, I'm leaning towards `write_out_train_split`, since even though one of the splits is called "test" it's still kind of part of the training process, and at least this clearly distinguishes the TFRecord splits from the true holdout test set. The docstring should also be updated to clarify what exactly a "train split" is and counteract the confusing-ness of one of the splits still being called "test".

**Review by kvablack — CHANGES_REQUESTED** [review]

File: ravenml/data/write_dataset.py (hunk 15 above)
Comment (AbhiDhir) [reply ravenml/data/write_dataset.py]: The `os.path.dirname` makes it so that it is the path to the imageset directory. However, when I change construct_all to return a dict instead, this logic won't be necessary any longer since the keys should be the image_id tuples in the correct format.

File: ravenml/data/write_dataset.py (hunk 2 above)
Comment (cSchubes) [reply ravenml/data/write_dataset.py:268]: @kevin3-black what do we want to do for this? is the added size of a kfold dataset so bad we should just ignore it altogether? otherwise I think enabling all datasets for possible kfold down the road would be best.

File: ravenml/data/helpers.py (hunk 3 above)
Comment (cSchubes) [reply ravenml/data/helpers.py:15]: I thought tag filtering and selecting only a subset of images were already part of the old Jigsaw, are those just being set up for new structure?

File: ravenml/data/write_dataset.py (hunk 17 above)
Comment (cSchubes) [reply ravenml/data/write_dataset.py]: I agree on this

File: ravenml/data/write_dataset.py (hunk 2 above)
Comment (kvablack) [reply ravenml/data/write_dataset.py:268]: I think the plan is still to add k-fold to the core, just not in this PR.

File: ravenml/data/helpers.py (hunk 3 above)
Comment (AbhiDhir) [reply ravenml/data/helpers.py:15]: The way the jigsaw filtering worked was that you can filter by certain tags (AND/OR between them) and then select a subset of those images. The feature I was referring to is the ability to just choose the number of images you want from each of your imagesets without having to deal with tags. The feature is finished and in the branch `better-filtering` which I will make a PR for after this one goes through.

**Review by kvablack — DISMISSED** [review]
LGTM! Great work on rewriting pretty much all of Jigsaw at this point. This base interface is solid, and I'm excited to see ravenML-data continue evolving and getting even more awesome features.

**Review by kvablack — APPROVED** [review]
